src/persist/src/storage.rs

```diff
@@ -211,6 +212,46 @@ pub trait Blob: BlobRead + Sized {
     async fn delete(&mut self, key: &str) -> Result<(), Error>;
 }
 
+/// An abstraction for a single arbitrarily-sized binary blob and a associated
+/// version number (sequence number).
+#[derive(Debug, Clone, PartialEq)]
+pub struct VersionedData {
+    /// The sequence number of the data.
+    pub seqno: SeqNo,
+    /// The data itself.
+    pub data: Vec<u8>,
+}
+
+/// An abstraction for [VersionedData] held in a location in persistent storage
+/// where the data are conditionally updated by version.
+///
+/// Users are expected to use this API with consistently increasing sequence numbers
+/// to allow multiple processes across multiple machines to agree to a total order
+/// of the evolution of the data.
+///
+/// TODO: Do we need to expose the evolution of this data across a series of versions?
+#[async_trait]
+pub trait Consensus: std::fmt::Debug {
+    /// Returns a recent version of `data`, and the corresponding sequence number, if
+    /// one exists at this location.
+    async fn head(&self, deadline: Instant) -> Result<Option<VersionedData>, Error>;
+
+    /// Update the [VersionedData] stored at this location to `new`, iff the current
+    /// sequence number is exactly `expected`.
+    ///
+    /// Returns a recent version and data from this location iff the current sequence
+    /// number does not equal `expected`.
+    ///
+    /// This data is initialized to None, and the first call to compare_and_swap needs to
+    /// happen with None as the expected value to set the state.
+    async fn compare_and_swap(
+        &mut self,
+        expected: Option<SeqNo>,
+        new: Option<VersionedData>,
+        deadline: Instant,
```

Comment:
nit: locally, I've been putting `deadline` as the first non-self param and it's been working out well. let's establish that as the convention